src/Data/Text/Utf16/Rope/Mixed.hs

```diff
@@ -78,26 +78,50 @@ import Text.Show (show)
 data Rope
   = Empty
   | Node
-    { _ropeLeft          :: !Rope
-    , _ropeMiddle        :: !TextLines
-    , _ropeRight         :: !Rope
-    , _ropeCharLen       :: !Word
-    , _ropeCharLenAsPos  :: !Char.Position
-    , _ropeUtf16Len      :: !Word
-    , _ropeUtf16LenAsPos :: !Utf16.Position
+    { _ropeLeft    :: !Rope
+    , _ropeMiddle  :: !TextLines
+    , _ropeRight   :: !Rope
+    , _ropeMetrics :: {-# UNPACK #-} !Metrics
     }
 
+data Metrics = Metrics
+  { _metricsCharLen       :: !Word
+  , _metricsCharLenAsPos  :: !Char.Position
+  , _metricsUtf16Len      :: !Word
+  , _metricsUtf16LenAsPos :: !Utf16.Position
+  }
+
 instance NFData Rope where
   rnf Empty = ()
   -- No need to deepseq strict fields, for which WHNF = NF
-  rnf (Node l _ r _ _ _ _) = rnf l `seq` rnf r
+  rnf (Node l _ r _) = rnf l `seq` rnf r
 
 instance Eq Rope where
   (==) = (==) `on` toLazyText
 
 instance Ord Rope where
   compare = compare `on` toLazyText
 
+instance Semigroup Metrics where
+  Metrics c1 p1 u1 up1 <> Metrics c2 p2 u2 up2 =
+    Metrics (c1 + c2) (p1 <> p2) (u1 + u2) (up1 <> up2)
```

Comment:
This might need `{-# INLINE #-}` just in case. Same in other modules.

Reply:
Done